README
rb3 is a tool for generating and managing configuration repositories.

Please report bugs to <sysdev@it.ox.ac.uk>, and send a blank email
to <sysdev-software-announce-subscribe@maillist.ox.ac.uk> to receive
notifications of updates.


debian/changelog
rb3 (1.42-1) UNRELEASED; urgency=medium

  * Team upload.
  * Bump compat from 4 to 10
  * Remove trailing whitespace from debian/changelog and debian/rules
  * Make this a non-native package
  * Add ${misc:Depends} to debian/control
  * Add Homepage: to debian/control
  * Remove custom field from debian/control
  * Switch debian/rules to dh
  * Update debian/copyright
  * Add debian/watch file
  * Patch for man page (man.patch)
  * Extend long description
  * Bump Standards-Version from 3.7.2 to 4.1.3 (no change required)
  * Patch for utf8

 -- Christopher Hoskin <mans0954@debian.org>  Mon, 22 Jan 2018 20:57:31 +0000

rb3 (1.42) stable; urgency=low

  [ Michael Howe ]
  * RB3::TemplateFunctions:
    - cache calls to the `hostname` function
  * RB3::CLI::Build:
    - remove automatic graphing on system build.  Closes srt#178525

  [ Kristian Kocher ]
  * output graphs with as_debug instead of as_canon (no pretty print) to avoid
    using dot in jessie
  * changed layout:
    - yml files have the same rank as the rb3 file that includes them
    - output files are generated from source files and rb3 files
    - source and output files are in a single column each
  * added template function to find interfaces from template toolkit files
  * don't die when reverse resolution fails for an IP address

  [ Dameon Wagner ]
  * Fix typos in rb3 POD.  Closes srt#189021.
  * Add subcommand "list-vms", for listing virtual machines that are
    configured on the host in rb3 (if any).
  * lib/RB3/CLI/FindInstalled.pm:
    - Fix query parameter string to hopefully catch all binary packages
      built from a searched-for source package.
    - Closes srt#214278.

 -- Michael Howe <michael.howe@it.ox.ac.uk>  Thu, 10 Sep 2015 08:45:17 +0100

rb3 (1.41.2) stable; urgency=low

  [ Kristian Kocher ]
  * Use file path and name as node names in graphs

 -- Kristian Kocher <kristian.kocher@it.ox.ac.uk>  Fri, 06 Jun 2014 09:06:14 +0100

rb3 (1.41.1) stable; urgency=low

  [ Kristian Kocher ]
  * Merged all graphing functions into one. The type of tree is controlled through options.
  * It is now possible to include template toolkit files in a "deep" tree.
  * Generated files and their source are now in different nodes in the tree.

  [ Michael Howe ]
  * Add support for ecdsa keys in RB3::CLI::UpdateSSHKeys.
  * Document (some of the) repository layout assumptions.

 -- Kristian Kocher <kristian.kocher@it.ox.ac.uk>  Thu, 06 Feb 2014 12:42:50 +0000

rb3 (1.40) stable; urgency=low

  * fixed a bug where data from different systems gets mangled when building
    multiple systems at the same time

 -- Kristian Kocher <kristian.kocher@it.ox.ac.uk>  Mon, 20 Jan 2014 17:44:26 +0000

rb3 (1.39) stable; urgency=low

  * Explicitly drop support for etch and lenny (perl <5.10.1) as they lack
    File::Path make_path support.

 -- Michael Howe <michael.howe@it.ox.ac.uk>  Mon, 20 Jan 2014 12:49:36 +0000

rb3 (1.38) stable; urgency=low

  [ Kristian Kocher ]
  * Integrating the graphing process into the main build process.

  [ Michael Howe ]
  * Make dump-params explicitly output in UTF-8.
  * New RB3::CLI::Graph module from Kristian Kocher (srt#131509)
  * Suggest xdot for displaying RB3::CLI::Graph-created graphs

  [ Dameon Wagner ]
  * bin/rb3
    -  Fix POD markup typo
    -  Add documentation about using a specific YAML file on a
       "per generated file" basis.

 -- Kristian Kocher <kristian.kocher@it.ox.ac.uk>  Fri, 17 Jan 2014 16:30:00 +0000

rb3 (1.37) unstable; urgency=low

  [ Tom Jones ]
  * Allow output location to be controlled via output_base and
    output_dir repository variables.  Enables better separation of
    input and output directories where required.
  * Add "hostname" package variable which defaults to the directory
    name but can be overridden with a repository variable also called
    "hostname".

  [ Dameon Wagner ]
  * Add subcommand "list-services", for listing the service interface
    addresses bound to the host.

  [ Tom Jones ]
  * Produce as UTF-8 from templates.  Authored by Michael Howe, ref #108540.

  [ Michael Howe ]
  * Make list-services recognize 'ip addr add' syntax, also use the
    Regexp::Common definition of IPv4 addresses.
  * list-services can take a hostname as well as system directory

  [ Dave Stewart ]
  * Fix why-path.  The diff  is quite ugly and I'd like to find a
    neater way of achieving the same thing.

 -- Dominic Hargreaves <dominic.hargreaves@it.ox.ac.uk>  Tue, 17 Sep 2013 11:43:01 +0100

rb3 (1.36) unstable; urgency=low

  * Per-repository file called .rb3repo instead of .rb3rc so as to
    avoid confusion when rb3 is being run from home directory.

 -- Tom Jones <tom.jones@oucs.ox.ac.uk>  Wed, 10 Oct 2012 16:34:38 +0100

rb3 (1.35) unstable; urgency=low

  * Allow strict to be set on a per-repository basis using a .rb3rc
    file in the repository, in the same format as ~/.rb3rc.
  * Make the find_netmask_in_table template function return zero instead
    of undef for a negative result.  It is an error in Template Toolkit's
    strict mode for a function to return undef.  We still agree with
    find_netmask_in_table's spec, which says it returns false if there's
    no match.
  * Don't emit "mkdir" messages in silent mode.
  * Document Template Toolkit VMethods that rb3 makes available to the
    templates.

 -- Tom Jones <tom.jones@oucs.ox.ac.uk>  Mon, 17 Sep 2012 13:51:35 +0100

rb3 (1.34) unstable; urgency=low

  * Print a nicer error report when a template fails to build.  It now
    contains the input and output file names, and the system for which
    it failed, in addition to the error message as previously.  The
    line of the template file causing the error still isn't printed.

 -- Tom Jones <tom.jones@oucs.ox.ac.uk>  Tue, 10 Jul 2012 21:34:20 +0100

rb3 (1.33) unstable; urgency=low

  * ipanyaddrs can now cope with CNAMEs.

 -- Tom Jones <tom.jones@oucs.ox.ac.uk>  Mon, 28 May 2012 17:50:09 +0100

rb3 (1.32) unstable; urgency=low


lib/RB3/CLI/Build.pm
#
# $HeadURL: https://svn.oucs.ox.ac.uk/sysdev/src/packages/r/rb3/tags/1.42/lib/RB3/CLI/Build.pm $
# $LastChangedRevision: 26904 $
# $LastChangedDate: 2015-09-10 08:49:03 +0100 (Thu, 10 Sep 2015) $
# $LastChangedBy: worc2070 $
#
package RB3::CLI::Build;

use strict;

use warnings FATAL => 'all';

use File::Basename qw( basename dirname );
use File::Path;
use File::Spec;
use File::Spec::Functions; # imports catfile
use IO::File;
use IO::Pipe;
#use List::MoreUtils qw(uniq);
use POSIX qw( SIGINT WIFSIGNALED WTERMSIG WIFEXITED WEXITSTATUS );
use RB3::Config;
use RB3::File;
use RB3::FileGenerator;
use RB3::CLI::Graph;
use YAML;

sub cmd_build {
    my $class = shift;
    my $app_config = shift;

    RB3::FileGenerator->DryRun( 1 )
          if $app_config->get( "dry-run" );

    RB3::FileGenerator->Quiet( 1 )
          if $app_config->quiet;

    RB3::FileGenerator->Silent( 1 )
          if $app_config->silent;

    RB3::FileGenerator->Strict( 1 )
          if $app_config->strict;

    my $max_jobno = $app_config->jobs - 1;

    # We'll rely on autovivification of arrayrefs inside @bkts later.
    my @bkts;
    my @children;

    for ( my $i = 0; $i < @_; $i++ ) {
        push @{ $bkts[ $i % $app_config->jobs ] }, $_[$i];
    }
    
    my $bkt;
    my @failhandles;

    FORK: for ( $bkt = 0; $bkt < $app_config->jobs; $bkt++ ) {

        my $failpipe = IO::Pipe->new;

        my $pid = fork;

        if( $pid ) {
            push @children, $pid;
            $failpipe->reader;
            push @failhandles, $failpipe;
            next FORK;
        }
        elsif( $pid == 0 ) {
            $failpipe->writer;

            $SIG{INT} = sub {
                File::Temp::cleanup();
                exit(1);
            };

            foreach my $sys ( @{ $bkts[$bkt] } ) {
                warn "Building configuration for $sys\n"
                    unless $app_config->silent;

                eval { build_host_config( $app_config, $sys ) } ;
                if( $@ ) {
                    print STDERR $@;
                    print $failpipe $sys . chr(0);
                }
            }
            exit 0;
        }
        elsif( not defined $pid ) {
            my $err = $!;
            reap_children([@failhandles]);
            die "fork failed: $err\n";
        }
    }

    reap_children([@failhandles]);
}

sub build_host_config {
    my ( $app_config, $hostdir ) = @_;

    unless ( -d $hostdir ) {
        warn "Skipping $hostdir (not a directory)\n";
        return;
    }

    my $rb3 = RB3::Config->new( { system_dir => $hostdir } );

    unless ( -r $rb3->get_rb3_path ) {
        warn "Skipping $hostdir (no config.rb3)\n";
        return;
    }

    $rb3->read_config();

    my $u_outbase = $rb3->get_repovars_list->template_vars->{"output_base"};
    my $u_outdir = $rb3->get_repovars_list->template_vars->{"output_dir"};

    my $outdir = defined($u_outdir) ? $u_outdir :
      defined($u_outbase)
      ? File::Spec->catfile($u_outbase, basename($rb3->get_system_dir))
      : $rb3->get_system_dir;


    my $fg = RB3::FileGenerator->new( { params     => $rb3->get_parameter_list,
                                        system_dir => $outdir,
    } );

    my $u_hostname = $rb3->get_repovars_list->template_vars->{"hostname"};
    my $hostname = defined($u_hostname)
      ? $u_hostname
      : basename($rb3->get_system_dir);

    foreach my $file ( @{ $rb3->get_file_list } ) {
        my $source = $file->get_source
            or next;

        eval {
            $fg->generate( 
                $source, $file->get_dest, $file->get_ctmeta_path, 
                $file->get_parameter_list, $file->get_component, $app_config,
                $hostname,
            );
        };
        if ($@) {
            die "Error generating file.\n"
                . "  source: $source\n"
                . "  dest:   " . $file->get_dest . "\n"
                . "  system: $hostdir\n"
                . "  err:    $@"
                ;
       }
    }

    write_configtool_meta($rb3, $fg, $app_config);
    write_configtool_manifests($rb3, $fg, $app_config);
}

sub write_configtool_meta {
    my ($rb3, $filegen, $app_config) = @_;

    my @metapaths;

    SET_METAPATHS: {
        my $metapaths
            = $rb3->get_repovars_list->template_vars->{"configtool.meta"};

        if (defined($metapaths)) {
            ref($metapaths) eq 'ARRAY' 
                or die "template var 'configtool.meta' must be a list\n";

            for my $mp (@$metapaths) {
                my %props = RB3::Config::decode_path_notation($mp);
                push @metapaths, $filegen->repopath(
                    $props{component}, $props{dest}
                );
            }      
        }
        else {
            @metapaths = (
                File::Spec->catfile($rb3->get_root_dir, 'etc', 'configtool.meta')
            );
        }
    }

    for my $path (@metapaths) {
        mkpath(dirname($path), !$app_config->silent, 0755);

        my $ofh = IO::File->new( $path, O_RDWR|O_CREAT|O_TRUNC, 0644 )
            or die "open $path for writing: $!";


        my @files = map { $_->[ 0 ] } sort { $a->[1] cmp $b->[1] }
            map { [ $_, $_->get_dest ] } @{ $rb3->get_file_list };

        foreach my $file ( @files ) {
            next if $file->get_owner eq RB3::File->DefaultOwner
                and $file->get_group eq RB3::File->DefaultGroup
                    and oct( $file->get_mode ) == oct( RB3::File->DefaultMode );
            $ofh->printf(
                "%s %s %s 0%o\n",
                (map { $file->$_ } qw( get_ctmeta_path get_owner get_group )), 
                oct( $file->get_mode )
            );
        }
    }
}

sub write_configtool_manifests {
    my ($rb3, $filegen, $app_config) = @_;

    my @mfpaths;

    SET_MANIFESTPATHS: {
        my $mfpaths
            = $rb3->get_repovars_list->template_vars->{"configtool.manifest"};

        if (defined($mfpaths)) {
            ref($mfpaths) eq 'ARRAY' 
                or die "template var 'configtool.manifest' must be a list\n";

            for my $mp (@$mfpaths) {
                my %props = RB3::Config::decode_path_notation($mp);
                push @mfpaths, $filegen->repopath(
                    $props{component}, $props{dest}
                );
            }
        }

        # If we want to generate the manifest in a default location for common
        # setups in future, the logic would go something like this (but
        # with proper component handling).

        # Reasoning: if there's a home component, manifest consumer
        # can probably write to their home directory but not /etc.  If
        # the only component is root, common case will be that the consumer 
        # of the manifest is running as root.
#       else {
             # Would need to uncomment "use List::MoreUtils" to get uniq.
#            my @comps = uniq(map { $_->get_component } @{$rb3->get_file_list};
#            if (grep { $_ eq 'home' } @comps) {
#                push @mfpaths, catfile(
#                    $rb3->get_root_dir, 'home/.configtool.manifest'
#                );
#            }
#            elsif (@comps == 1 and $comps[0] eq 'root') {
#                push @mfpaths, catfile(
#                    $rb3->get_root_dir, 'root/etc/configtool.manifest'
#                );
#            }
#        }
    }

    return unless @mfpaths;

    my @manifest_data;

    my @files = map { $_->[ 0 ] } sort { $a->[1] cmp $b->[1] }
            map { [ $_, $_->get_dest ] } @{ $rb3->get_file_list };

    foreach my $file ( @files ) {
        next if !$file->get_owner_explicitly_set
            and !$file->get_group_explicitly_set
            and !$file->get_mode_explicitly_set
            and !scalar(%{$file->get_extras});

        my %mfentry = (path => $file->get_ctmeta_path);
        $file->get_owner_explicitly_set and $mfentry{owner} = $file->get_owner;
        $file->get_group_explicitly_set and $mfentry{group} = $file->get_group;
        $file->get_mode_explicitly_set and $mfentry{mode} = $file->get_mode;

        %mfentry = (%mfentry, %{$file->get_extras});

        push @manifest_data, \%mfentry;
    }
    
    for my $path ( @mfpaths ) {
        mkpath(dirname($path), !$app_config->silent, 0755);

        my $ofh = IO::File->new( $path, O_RDWR|O_CREAT|O_TRUNC, 0644 )
            or die "open $path for writing: $!";

        print $ofh YAML::Dump(\@manifest_data);
    }
}

sub reap_children {
    my ($failpipes) = @_;

    my $ex = 0;

    my @failed_systems;

    # unfortunately not settable on a per-filehandle scope
    $/ = chr(0);

    for my $failpipe ( @$failpipes ) {
        while( my $fail = <$failpipe> ) {
            chomp $fail;
            push @failed_systems, $fail;
        }
    }

    $/ = "\n";

    if( @failed_systems ) {
        warn(
            "system dirs failing to build:\n"
                . join("", map { "$_\n" } @failed_systems)
        );
        $ex = 1;
    }

    my $errors = 0;

    CHILD: while( my $child = wait ) {
        $child == -1 and last CHILD;

        if( WIFSIGNALED( $? ) ) {
            warn "child $child died with signal " . WTERMSIG( $? ) . "\n";
            $errors++;
        }
        elsif( WIFEXITED( $? ) and WEXITSTATUS( $? ) != 0 ) {
            warn "child $child exited with value " . WEXITSTATUS( $? ) . "\n";
            $errors++;
        }
    }

    if( $errors ) {
        warn "children not exiting successfully: $errors\n";
        $ex = 1;
    }

    $ex and exit( $ex );
}

1;

__END__
